### ISSUE TITLE
Add multi retryRef to action

### ISSUE THREAD

**Issue by lsytj0413** [serverlessworkflow/specification#681]
**What would you like to be added**:

In current version, there is only one retry strategy can be assign to action:
```
"actions": [
            {
              "functionRef": "StorePatient",
              "retryRef": "ServicesNotAvailableRetryStrategy",
              "retryableErrors": ["ServiceNotAvailable"]
            }]

 "retries": [
    {
      "name": "ServicesNotAvailableRetryStrategy",
      "delay": "PT3S",
      "maxAttempts": 10
    }
  ]
```

If a action can reference multi retry strategy it will be useful.

**Why is this needed**:

+ amazon states language support multi retry strategy https://states-language.net/spec.html#errors
+ If the action failed, user may want use different retry strategy, for example:
  + delay 1s if the error is short-term（ex: FlowLimit）
  + delay 5s if the error is long-term（ex: database is not available）

**Comment by ricardozanini** [comment]
This is interesting. I think it can be useful to add the retriable error in the retry definition as Amazon States Language does, or it will be difficult to understand the mapping of two arrays. Technically, it will be the same, but it will improve understanding.

Can be something like this:

```json
{
   "actions":[
      {
         "functionRef":"StorePatient",
         "retryRefs":[
            "ServicesNotAvailableRetryStrategy",
            "NetworkIssue"
         ]
      }
   ],
   "retries":[
      {
         "name":"ServicesNotAvailableRetryStrategy",
         "delay":"PT3S",
         "maxAttempts":10,
         "errors":[
            "ServiceNotAvailable"
         ]
      },
      {
         "name":"NetworkIssue",
         "delay":"PT10S",
         "maxAttempts":3,
         "errors":[
            "ConnectionFailed"
         ]
      }
   ]
}

```

**Comment by lsytj0413** [comment]
> 

Currently，the `retryableErrors` is in Action's definition.

**Comment by tsurdilo** [comment]
I think this would be nice feature to add (even tho possibly very hard to implement). Also I think would not be very useful for idempotent actions.

Looking at the mentioned AWS samples:

```
"Retry" : [
    {
      "ErrorEquals": [ "States.Timeout" ],
      "MaxAttempts": 0
    },
    {
      "ErrorEquals": [ "States.ALL" ]
    }
]
```

I think this can be implemented already by setting States.Timeout as a non-retryable error.

```
"Retry": [
    {
      "ErrorEquals": [ "ErrorA", "ErrorB" ],
      "IntervalSeconds": 1,
      "BackoffRate": 2,
      "MaxAttempts": 2
    },
    {
      "ErrorEquals": [ "ErrorC" ],
      "IntervalSeconds": 5
    }
  ]
```

This is what we are talking about I think. I'd like to know more on how they are handling this when different errors have different maxAttempt values. Think this can run into a number of edge cases that would need to be clearly described.

**Comment by lsytj0413** [comment]
Can you give more example on `not useful for idempotent actions`?
> I think this would be nice feature to add (even tho possibly very hard to implement). Also I think would not be very useful for idempotent actions.

imo the specification should define how to handing maxAttempt when different errors, like use the value based on error. For example:

1. we have an action named `CreateDisk`, and the action maybe failed with `FlowLimit`，`InsufficientResource`
2. When error is `InsufficientResource`, we only retry for 3 times
3. When error is `FlowLimit`, we retry for 30-times

**Comment by tsurdilo** [comment]
Yes I updated some text in my answer above (please look again). 

The retry with this is tied to an error type. In this case we can either assume that if the first invocation raises FlowLimit its retried up to 30 times regardless of errors raised by consequent retries.
Or you would re-calculate the retry policy if the second retry throws InsufficientResource.

I think this is doable iff we go with the first assumption. wdyt.

**Comment by lsytj0413** [comment]
I don't known how AWS does it, but in my company's workflow engine（base on ASL），it will use the value based on error and record the retry history. For example: the CreateDisk may retry upon 30 times when error is mixed with FlowLimit & InsufficientResource.

> I think this is doable iff we go with the first assumption (which i think is what aws does too but not sure, let me know if you know). wdyt.

**Comment by tsurdilo** [comment]
Just to add as a general info thing, imo it is an anti-pattern to limit retries via MaxAttempts.
For most use cases you should limit your overall action retries using state timeout (thats what its designed for at least imo :) )

**Comment by tsurdilo** [comment]
> but MaxAttempt is useful when we want fast-to-fail if we support multi retry strategy.

I think this really depends on the action timeout (single action execution limit). If you set the action timeout to a large value, and your action is "stuck" for minutes or hours you cannot fail fast even with just 2 retries. 

The best way to deal with your use cases is to have your actions "heartbeat" meaning report progress and have a heartbeat timeout set to a very small value within which the action has the report its progress/heartbeat. 

This however idk if we can impl in our DSL as not sure we can enforce heartbeating to all runtimes.

**Comment by tsurdilo** [comment]
Overall +1 on adding this, just would like to maybe get one or two examples on how this would look like if its done via the mentioned dynamic retry policy calculations on each retry. Thanks!

**Comment by lsytj0413** [comment]
> I think this really depends on the action timeout (single action execution limit). If you set the action timeout to a large value, and your action is "stuck" for minutes or hours you cannot fail fast even with just 2 retries.

Base on the exmaple, we want 3 MaxAttempts and Interval 5s when `InsufficientResource`，30 MaxAttempts and Interval 1s when `FlowLimit`，we can only set the ActionTimeout to 30s=max(15s, 30s)，but if the only error is `InsufficientResource` we can stuck only 15s（less than 30s）.

**Comment by lsytj0413** [comment]
@tsurdilo  What's the next step should i do for this？is an PR ok？

**Comment by cdavernas** [comment]
Closed as resolved by 0.10, and therefore as part of #843
